### ISSUE TITLE
The rollback function no longer exists

### ISSUE THREAD

**Issue by benbankes** [tannerlinsley/react-query#664]
**Describe the bug**
After upgrading from 2.2.3 to 2.4.5, the `rollback()` function no longer exists for optimistic UI updates.  This change is not backward compatible.

**To Reproduce**
1. Use the first, optimistic UI update mechanism described in the [readme](https://github.com/tannerlinsley/react-query#updating-a-list-of-todos-when-adding-a-new-todo)
2. See the error message `Uncaught (in promise) TypeError: rollback is not a function` in the console.

**Expected behavior**
I expected the rollback function to be available.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Browser Chrome
 - Version 83

**Additional context**
The function signature for the `onSuccess` handler was previously:
`(err, newTodo, rollback) => void`
Now it is:
`(err, newTodo, snapshotValue) => void | Promise<void>`

If this change was intentional, the simple workaround for 2.4.5 is to replace a call to:
`rollback()`
with
`queryCache.setQueryData('todos, snapshotValue);`
If semantic versioning is being followed, it might make sense to revert this change, and defer the backward compatibility break to version 3.

**Comment by benbankes** [comment]
My suggested workaround will not work, because it causes the `error` to be overwritten after a subsequent render.

**Comment by denisborovikov** [comment]
A rollback was always a value returned from the onMutate handler. In the example, it's a function, but it's doesn't have to be. 

But anyway, according to [docs](https://github.com/tannerlinsley/react-query#options-3) onSuccess handler doesn't receive a snapshotValue, because it doesn't make sense to roll back if there were no errors during the mutation. In case of success, the cache will be updated with the new value from the server.

**Comment by benbankes** [comment]
Aha!  I figured it out.

**Background**
I composed several react-query hooks into a custom hook, and I used that custom hook both in one component, and in another component, nested within the original component.

**My Incorrect Assumption**
Since network requests were not duplicated by doubly using this hook, I incorrectly assumed, both that the internal behavior of the custom hook would not be duplicated, and that the values returned from the hook would not be different between components.

**Documentation**
However, the official [Custom Hooks] documentation states otherwise:

>**Do two components using the same Hook share state?**
>No. Custom Hooks are a mechanism to reuse stateful logic (such as setting up a subscription and remembering the current value), but every time you use a custom Hook, all state and effects inside of it are fully isolated.

I guess I still have more to learn! 😄
